<h1 align=center style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
Mobile Price
</font>
</h1>
<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    We want to guess the price range of mobile phones using machine learning algorithms such as artificial neural networks (ANN), k nearest neighbors (KNN), support vector machain (SVM) etc...
</font>
</p>
<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    First, we have to normalize the data and split the training and test data, and in the next step, we implement the algorithms on it.
</font>
</p>

<h2 align=left style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
Librarys
</font>
</h2>
<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Import required libraries : 
</font>
</p>

In [25]:
import numpy as np
import pandas as pd
from sklearn.metrics import accuracy_score, precision_score, recall_score

<h2 align=left style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
Read data
</font>
</h2>
<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    From mobile_prices file read dataset :
</font>
</p>

In [1]:
df = pd.read_csv('mobile_prices.csv')
df.head()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  m_dep  \
0            842     0          2.2         0   1       0           7    0.6   
1           1021     1          0.5         1   0       1          53    0.7   
2            563     1          0.5         1   2       1          41    0.9   
3            615     1          2.5         0   0       0          10    0.8   
4           1821     1          1.2         0  13       1          44    0.6   

   mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0        188        2  ...         20       756  2549     9     7         19   
1        136        3  ...        905      1988  2631    17     3          7   
2        145        5  ...       1263      1716  2603    11     2          9   
3        131        6  ...       1216      1786  2769    16     8         11   
4        141        2  ...       1208      1212  1411     8     2         15   

   three_g  touch_screen  wifi  price_range  
0        0             0     1            1  
1        1             1     0            2  
2        1             1     0            2  
3        1             0     0            2  
4        1             1     0            1  

[5 rows x 21 columns]

<h2 align=left style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
Labels and Features
</font>
</h2>
<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Get features as X and labels as y<br>
    X : Pregnancies to Age<br>
    y : contain outcome
</font>
</p>

In [17]:
X = df.iloc[:, :-1].values
y = df.iloc[:, -1].values

<h2 align=left style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
Normalize
</font>
</h2>
<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    We equate all data between 0 and 1 by scaler from sklearn :
</font>
</p>

In [18]:
from sklearn.preprocessing import StandardScaler

X = StandardScaler().fit_transform(X=X)

X

array([[-0.90259726, -0.9900495 ,  0.83077942, ..., -1.78686097,
        -1.00601811,  0.98609664],
       [-0.49513857,  1.0100505 , -1.2530642 , ...,  0.55964063,
         0.99401789, -1.01409939],
       [-1.5376865 ,  1.0100505 , -1.2530642 , ...,  0.55964063,
         0.99401789, -1.01409939],
       ...,
       [ 1.53077336, -0.9900495 , -0.76274805, ...,  0.55964063,
         0.99401789, -1.01409939],
       [ 0.62252745, -0.9900495 , -0.76274805, ...,  0.55964063,
         0.99401789,  0.98609664],
       [-1.65833069,  1.0100505 ,  0.58562134, ...,  0.55964063,
         0.99401789,  0.98609664]])

<h2 align=left style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
Train and Test
</font>
</h2>
<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Split of train and test data with train_test_split from sklearn
</font>
</p>

In [21]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train , y_test = train_test_split(X, y, test_size=0.2)

<h2 align=left style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
Naive bayes
</font>
</h2>

In [44]:
from sklearn.naive_bayes import GaussianNB

naive_model = GaussianNB()
naive_model.fit(X=X_train,y=y_train)

GaussianNB()

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Test model : 
</font>
</p>

In [48]:
y_pred_test_naive = naive_model.predict(X_test)

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Calclate accuracy :
</font>
</p>

In [49]:
accuracy_naive = accuracy_score(y_pred=y_pred_test_naive, y_true=y_test)
accuracy_naive

0.825

<h2 align=left style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
Decision Tree
</font>
</h2>

In [88]:
from sklearn.tree import DecisionTreeClassifier

model_tree = DecisionTreeClassifier(max_depth=32, min_samples_split=2, min_samples_leaf=3, max_features=14)

model_tree.fit(X=X_train,y=y_train)

DecisionTreeClassifier(max_depth=32, max_features=14, min_samples_leaf=3)

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Test model : 
</font>
</p>

In [89]:
y_pred_test_tree = model_tree.predict(X_test)

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Calclate accuracy :
</font>
</p>

In [90]:
accuracy_tree = accuracy_score(y_pred=y_pred_test_tree, y_true=y_test)
accuracy_tree

0.8575

<h2 align=left style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
Random Forest
</font>
</h2>

In [140]:
from sklearn.ensemble import RandomForestClassifier

model_forest = RandomForestClassifier(n_estimators=100, max_depth=32, min_samples_split=2, min_samples_leaf=2, max_features=16)

model_forest.fit(X=X_train, y=y_train)

RandomForestClassifier(max_depth=32, max_features=16, min_samples_leaf=2)

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Test model : 
</font>
</p>

In [141]:
y_pred_test_forest = model_forest.predict(X_test)

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Calclate accuracy :
</font>
</p>

In [142]:
accuracy_forest = accuracy_score(y_pred=y_pred_test_forest, y_true=y_test)
accuracy_forest

0.91

<h2 align=left style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
Support Vector Machian (SVM)
</font>
</h2>

In [146]:
from sklearn import svm

model_svm = svm.SVC(kernel='linear')

model_svm.fit(X=X_train,y=y_train)

SVC(kernel='linear')

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Test model : 
</font>
</p>

In [147]:
y_pred_test_svm = model_svm.predict(X_test)

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Calclate accuracy :
</font>
</p>

In [149]:
accuracy_svm = accuracy_score(y_pred=y_pred_test_svm, y_true=y_test)
accuracy_svm

0.965

<h2 align=left style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
k nearest neighbors (KNN)
</font>
</h2>

In [186]:
from sklearn.neighbors import KNeighborsClassifier

model_knn = KNeighborsClassifier(n_neighbors=34)

model_knn.fit(X=X_train,y=y_train)

KNeighborsClassifier(n_neighbors=34)

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Test model : 
</font>
</p>

In [187]:
y_pred_test_knn = model_knn.predict(X_test)

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Calclate accuracy :
</font>
</p>

In [188]:
accuracy_knn = accuracy_score(y_pred=y_pred_test_knn, y_true=y_test)
accuracy_knn

0.6275

<h2 align=left style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
Logistic Regression
</font>
</h2>

In [189]:
from sklearn.linear_model import LogisticRegression

model_logistic = LogisticRegression()

model_logistic.fit(X=X_train, y=y_train)

LogisticRegression()

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Test model : 
</font>
</p>

In [190]:
y_pred_test_logistic = model_logistic.predict(X_test)

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Calclate accuracy :
</font>
</p>

In [191]:
accuracy_logistic = accuracy_score(y_pred=y_pred_test_logistic, y_true=y_test)
accuracy_logistic

0.97

<h2 align=left style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
Artificial Neural Networks (ANN)
</font>
</h2>

In [211]:
from sklearn.neural_network import MLPClassifier

model_MLP = MLPClassifier(hidden_layer_sizes=130,max_iter=400)

model_MLP.fit(X=X_train, y=y_train)

MLPClassifier(hidden_layer_sizes=130, max_iter=400)

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Test model : 
</font>
</p>

In [212]:
y_pred_test_MLP = model_MLP.predict(X_test)

<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    Calclate accuracy :
</font>
</p>

In [213]:
accuracy_MLP = accuracy_score(y_pred=y_pred_test_MLP, y_true=y_test)
accuracy_MLP

0.9325

<h1 align=center style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
Result
</font>
</h1>
<p dir=rtl style="direction: ltr ;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    We reviewed and implemented seven machine learning algorithms on the mobile price dataset and optimized them as much as possible by determining their parameters. The logistic regression algorithm had the highest accuracy with 97% and the k nearest neighbors (KNN) algorithm had the lowest accuracy with 62%.
</font>
</p>